### DISCUSSION TITLE
Expected flow for expiring credentials

### DISCUSSION THREAD

**Discussion by MrDoowie** [Automattic/mongoose#12698] [Q&A] [answered]
We're trying to use the MongoDB AWS authentication system for securing our connection (https://github.com/mongodb/specifications/blob/master/source/auth/auth.rst#mongodb-aws).
The credentials we provide to it are retrieved through the '@aws-sdk/client-sts' package (https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/clients/client-sts/index.html).

Our initial connection is working fine, but we're trying to work out how to refresh these credentials before they expire. Currently we can't see any other way other than:

- Create a new database connection
- Create new models using the connection
- Replace all references of existing models across the code base to use the new models
- Retire the old connection

This gets complicated fast. The initial model references are retrieved after the initial database connection is established in different locations, so replacing them all with newer instances isn't as straight forward as it sounds. Furthermore, if there are db operations that are in progress at the time of establishing a new connection (for example a transaction), it would need to be aware of this, and keep using the old connection until ready to start using the new connection.

Is there a way of refreshing the connection with new credentials, without tearing everything down and re-creating the connection and models? (Refreshing the existing AWS credential expiry time isn't an option, as there is a maximum time these credentials can be refreshed).

Thanks,
Michael

**Comment by vkarpov15** [comment, marked as answer]
There currently isn't any way to easily refresh credentials. You would need to close the old connection and create a new one. We'll open an issue to investigate how we can better support this.
